Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,739 @@
+  public static func * (lhs: BigInt, rhs: BigInt) -> BigInt {
+    let lhsIsNeg = lhs.words[lhs.words.endIndex - 1] > Int.max
+    let rhsIsNeg = rhs.words[rhs.words.endIndex - 1] > Int.max
+
+    let lhsWords = lhsIsNeg ? (-lhs).words : lhs.words
+    let rhsWords = rhsIsNeg ? (-rhs).words : rhs.words
+
+    let count = lhsWords.count + rhsWords.count + 1
+    var newWords = Words(repeating: 0, count: count)
+
+    for i in 0 ..< rhsWords.count {
+      var carry: UInt = 0
+      var digit: UInt = 0
+
+      var lastJ: Int = 0
+      for j in i ..< (lhsWords.count + i) {
+        var (high, low) = rhsWords[i].multipliedFullWidth(by: lhsWords[j - i])
+        var isOverflow: Bool
+        (digit, isOverflow) = low.addingReportingOverflow(newWords[j])
+        if isOverflow {
+          high += 1
+        }
+
+        (digit, isOverflow) = digit.addingReportingOverflow(carry)
+        if isOverflow {
+          high += 1
+        }
+
+        carry = high
+        newWords[j] = digit
+        lastJ = j
+      }
+
+      if carry != 0 {
+        let isOverflow: Bool
+        (digit, isOverflow) = newWords[lastJ + 1].addingReportingOverflow(carry)
+        if isOverflow {
+          carry = 1
+        }
+        newWords[lastJ + 1] = digit
+      }
+    }
+
+    for i in stride(from: count - 1, through: 1, by: -1) {
+      if newWords[i] == 0, newWords[i - 1] <= Int.max {
+        newWords.removeLast()
+      } else {
+        break
+      }
+    }
+
+    if lhsIsNeg || rhsIsNeg, !(lhsIsNeg && rhsIsNeg) {
+      return -BigInt(_uncheckedWords: newWords)
+    }
+
+    return BigInt(_uncheckedWords: newWords)
+  }
+
+  @inlinable
+  public static func *= (lhs: inout BigInt, rhs: BigInt) {
+    lhs = lhs * rhs
+  }
+}
+
+extension BigInt: SignedNumeric {
+
+  public mutating func negate() {
+    var isOverflow = true
+    for i in 0 ..< words.count {
+      if isOverflow {
+        (words[i], isOverflow) = (~words[i]).addingReportingOverflow(1)
+      } else {
+        words[i] = ~words[i]
+      }
+    }
+
+    BigInt._dropExcessWords(words: &words)
+  }
+
+  @inlinable
+  public static prefix func - (x: BigInt) -> BigInt {
+    var result = x
+    result.negate()
+    return result
+  }
+}
+
+extension BigInt: BinaryInteger {
+
+  public init?<T>(exactly source: T) where T: BinaryFloatingPoint {
+    guard source.isFinite, source == source.rounded(.towardZero) else {
+      return nil
+    }
+    self.init(source)
+  }
+
+  public init<T>(_ source: T) where T: BinaryFloatingPoint {
+    precondition(
+      source.isFinite,
+      """
+      \(type(of: source)) value cannot be converted to BigInt because it is \
+      either infinite or NaN
+      """)
+
+    let isNegative = source < 0.0
+    var float = isNegative ? -source : source
+
+    if let _ = UInt(exactly: T.greatestFiniteMagnitude) {
+      words = [UInt(float)]
+    } else {
+      var words = Words()
+      let radix = T(sign: .plus, exponent: T.Exponent(UInt.bitWidth), significand: 1)
```

Comment:
Float16 has a 5-bit exponent (-14 ... 15) so will `T.Exponent(UInt.bitWidth)` be rejected?

Reply:
It appears it would. There is, AFAIK, currently no Float16 type in Swift; however, there probably will be at some point in the future (it's supported in clang as `_Float16` after all). I've got an experimental branch where I'm implementing one (badly, just for testing) using clang's `_Float16` under the covers, but it's not really working on its own yet haha. Once I get that done, I can at least test this on that branch and figure out a solution (probably to bring back your bit-twiddling solution for `BinaryFloatingPoint` types with bounded `significandBitCount`).

Reply:
Answer is: it works fine! And the main reason is it never hits this code in the Float16 case, because `Float16.greatestFiniteMagnitude` is 65504, which obviously fits exactly in a `UInt`, so the fast path happens. I bet one could prove that any type whose exponent is less than or equal to 6 bits (5 bits in the 32-bit `UInt` case) will always have a `greatestFiniteMagnitude` that fits exactly in `UInt`.